include/storage/fbw.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <storage/flash_map.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct fbw_ctx {
+	u8_t *buf;
+	size_t buf_len;
+	const struct flash_area *flash_area;
+	size_t bytes_written;
+	u16_t buf_bytes;
+#ifdef CONFIG_IMG_ERASE_PROGRESSIVELY
+	off_t off_last;
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * @param ctx context to be initialized
+ * @param area_id Area ID to write to
+ * @param offset Offset within area to write to
+ *
+ * @return  0 on success, negative errno code on fail
+ */
+int fbw_init(struct fbw_ctx *ctx, u8_t *buf, size_t buf_len, u8_t area_id,
+	     size_t offset);
+
+/**
+ * @brief Read number of bytes written to the flash.
+ *
+ * @param ctx context
+ *
+ * @return Number of bytes written to the image flash.
+ */
+size_t fbw_bytes_written(struct fbw_ctx *ctx);
+
+/**
+ * @brief  Process input buffers to be written to the image slot 1. flash
+ * memory in single blocks. Will store remainder between calls.
+ *
+ * A final call to this function with flush set to true
+ * will write out the remaining block buffer to flash. Since flash is written to
+ * in blocks, the contents of flash from the last byte written up to the next
+ * multiple of CONFIG_IMG_BLOCK_BUF_SIZE is padded with 0xff.
```

Comment:
Now that this API is being generalized, does the padding value need to be a parameter to the function or left up to the flash device instead of hard-coded to 0xff?

Reply:
That makes sense, especially as this can be used against any type of flash device.

Reply:
I'm not padding the buffer anymore, since the pages are erased anyway this does not makes sense IMO.